R/CompDbSource.R

```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
-			out[[i]] <- lapply(1:length(adduct),function (i) {
-				neutralPrecursorMass <- MetaboCoreUtils::mz2mass(mz, adduct[i])
-				restQuery <- sapply(slotNames(param)[-which(slotNames(param) %in% c("validAdducts","THRESHFUN"))], function(x) { 
-					slot(param,x)
-				})
-				restQuery <- c(
-					restQuery,	
-					"neutralprecursormass" = as.character(neutralPrecursorMass),
-					"precursoriontype" = as.character(adduct),
-					"peakliststring" = peakListString
+				# next()
+                peakListString <- ""
+			} else
+                peakListString <- .peaks_to_metfrag_string(peaks)
+			pmz <- qi$precursorMz
+			adduct <- unlist(qi$precursorAdduct)
+			out[[i]] <- do.call(rbind, lapply(adduct, function(a) {
+				query <- c(
+					restQuery,
+					neutralprecursormass = as.character(mz2mass(pmz, a)),
+					precursoriontype = a,
+					peakliststring = peakListString
 				)
-				
-				mfRequest <- POST(url = target@url,config = target@config,body = restQuery,encode = "json",verbose())
-						
-				## get the http call status, plus results(if exist): 
-				callStatus <- .chk_rest_response (mfRequest)
+				mfRequest <- POST(url = target@url, config = target@config,
+                                  body = query, encode = "json", handle)
+				## get the http call status, plus results(if exist):
+				callStatus <- .chk_rest_response(mfRequest)
 				if (callStatus$status == "SUCCESS" ) {
-					mfCandidates <- read.csv(text = content(GET(callStatus$rest_url))) 
-				} else {				
-					mfCandidates = NULL
-					## Warning msg?
-				}
-				## return a list of anntation results table and server staus code
-				list(
-					mfResults = mfCandidates,
-					mfServerStatus = callStatus$status
-				)
-			})
-			names(out[[i]]) <- adduct
+					res <- read.csv(text = content(GET(callStatus$rest_url)))
+                    cbind(res, data.frame(query_idx = rep(i, nrow(res)),
+                                          adduct = rep(a, nrow(res))))
+                } else {
+                    warning("No result for query spectrum [", i, "], adduct \"",
+                            a, "\". Response from server was: ",
+                            callStatus$status)
+                    ## Suggestion: don't report anything if MetFrag did not
+                    ## return any result (no matter why). Alternative would be
+                    ## to report `NA` for all fields, which might however be
+                    ## misleading to the user (because a *match* would be
+                    ## reported).
+                    data.frame()
+                }
+			}))
 			Sys.sleep(1)
 		}
-		out	
-		#' ToDo:
-		#' Convert nested list to the MatchedSpectra Object?
-		#' ...
-		#' ...		
+		out <- do.call(rbind, out)
+        mtches <- data.frame(query_idx = out$query_idx,
+                             target_idx = seq_len(nrow(out)),
+                             score = out$Score,
+                             adduct = out$adduct)
+        Matched(query = query,
```

Comment:
Result gets returned as a `Matched` object with `query` being the input `Spectra` and `target` the result from MetFrag

Reply:
Will be happy to see that work!